File: frontend/src/types/scriptLibrary.ts
/**
 * Script Library Types
 */

export interface ScriptSeedCharacter {
  name: string;
  age: number;
  surface_identity: string;
  true_identity: string;
  profession: string;
}

export interface ScriptSeedContrast {
  surface: string;
  truth: string;
  hook: string;
}

export interface ScriptSeedProgression {
  start: string;
  build: string;
  climax: string;
  resolve: string;
}

export interface ScriptSeedKeyNode {
  name: string;
  description: string;
  trigger: string;
}

export interface ScriptSeedEndings {
  good: string;
  neutral: string;
  bad: string;
  secret: string;
}

export interface ScriptSeed {
  character: ScriptSeedCharacter;
  contrast: ScriptSeedContrast;
  progression: ScriptSeedProgression;
  key_nodes: ScriptSeedKeyNode[];
  endings: ScriptSeedEndings;
}

export interface ScriptLibrary {
  id: string;
  title: string;
  title_en?: string;
  summary?: string;
  emotion_tones: string[];
  relation_types: string[];
  contrast_types: string[];
  era?: string;
  gender_target?: string;
  character_gender?: string;
  profession?: string;
  length?: string;
  age_rating?: string;
  contrast_surface?: string;
  contrast_truth?: string;
  contrast_hook?: string;
  script_seed?: ScriptSeed;
  full_script?: Record<string, unknown>;
  popularity: number;
  status: 'draft' | 'published' | 'archived';
  created_at: string;
  updated_at: string;
}

export interface ScriptLibraryListResponse {
  items: ScriptLibrary[];
  total: number;
  page: number;
  page_size: number;
}

export interface ScriptTag {
  id: string;
  category: string;
  name: string;
  name_en?: string;
  description?: string;
  examples: string[];
  parent_id?: string;
}

export interface ScriptTagsByCategory {
  emotion_tones: ScriptTag[];
  relation_types: ScriptTag[];
  contrast_types: ScriptTag[];
  eras: ScriptTag[];
  professions: ScriptTag[];
  gender_targets: ScriptTag[];
  character_genders: ScriptTag[];
  lengths: ScriptTag[];
  age_ratings: ScriptTag[];
}

export interface ScriptLibraryFilter {
  emotion_tones?: string[];
  relation_types?: string[];
  contrast_types?: string[];
  era?: string;
  gender_target?: string;
  character_gender?: string;
  profession?: string;
  age_rating?: string;
  length?: string;
  search?: string;
  status?: string;
  page?: number;
  page_size?: number;
}

export interface ScriptLibraryCreate {
  title: string;
  title_en?: string;
  summary?: string;
  emotion_tones?: string[];
  relation_types?: string[];
  contrast_types?: string[];
  era?: string;
  gender_target?: string;
  character_gender?: string;
  profession?: string;
  length?: string;
  age_rating?: string;
  contrast_surface?: string;
  contrast_truth?: string;
  contrast_hook?: string;
  script_seed?: ScriptSeed;
  full_script?: Record<string, unknown>;
}

export interface ScriptLibraryUpdate {
  title?: string;
  title_en?: string;
  summary?: string;
  emotion_tones?: string[];
  relation_types?: string[];
  contrast_types?: string[];
  era?: string;
  gender_target?: string;
  character_gender?: string;
  profession?: string;
  length?: string;
  age_rating?: string;
  contrast_surface?: string;
  contrast_truth?: string;
  contrast_hook?: string;
  script_seed?: ScriptSeed;
  full_script?: Record<string, unknown>;
  status?: 'draft' | 'published' | 'archived';
}

File: backend/app/models/script_library.py
"""
Script Library Data Models
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from enum import Enum


class ScriptLibraryStatus(str, Enum):
    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class ScriptSeedCharacter(BaseModel):
    name: str
    age: int
    surface_identity: str
    true_identity: str
    profession: str


class ScriptSeedContrast(BaseModel):
    surface: str
    truth: str
    hook: str


class ScriptSeedProgression(BaseModel):
    start: str
    build: str
    climax: str
    resolve: str


class ScriptSeedKeyNode(BaseModel):
    name: str
    description: str
    trigger: str


class ScriptSeedEndings(BaseModel):
    good: str
    neutral: str
    bad: str
    secret: str


class ScriptSeed(BaseModel):
    character: ScriptSeedCharacter
    contrast: ScriptSeedContrast
    progression: ScriptSeedProgression
    key_nodes: List[ScriptSeedKeyNode]
    endings: ScriptSeedEndings


class ScriptLibraryBase(BaseModel):
    title: str
    title_en: Optional[str] = None
    summary: Optional[str] = None
    
    emotion_tones: List[str] = []
    relation_types: List[str] = []
    contrast_types: List[str] = []
    era: Optional[str] = None
    gender_target: Optional[str] = None
    character_gender: Optional[str] = None
    profession: Optional[str] = None
    length: Optional[str] = None
    age_rating: Optional[str] = None
    
    contrast_surface: Optional[str] = None
    contrast_truth: Optional[str] = None
    contrast_hook: Optional[str] = None
    
    script_seed: Optional[ScriptSeed] = None
    full_script: Optional[Dict[str, Any]] = None


class ScriptLibraryCreate(ScriptLibraryBase):
    pass


class ScriptLibraryUpdate(BaseModel):
    title: Optional[str] = None
    title_en: Optional[str] = None
    summary: Optional[str] = None
    emotion_tones: Optional[List[str]] = None
    relation_types: Optional[List[str]] = None
    contrast_types: Optional[List[str]] = None
    era: Optional[str] = None
    gender_target: Optional[str] = None
    character_gender: Optional[str] = None
    profession: Optional[str] = None
    length: Optional[str] = None
    age_rating: Optional[str] = None
    contrast_surface: Optional[str] = None
    contrast_truth: Optional[str] = None
    contrast_hook: Optional[str] = None
    script_seed: Optional[ScriptSeed] = None
    full_script: Optional[Dict[str, Any]] = None
    status: Optional[ScriptLibraryStatus] = None


class ScriptLibrary(ScriptLibraryBase):
    id: str
    popularity: int = 0
    status: ScriptLibraryStatus = ScriptLibraryStatus.DRAFT
    created_at: str
    updated_at: str
    
    class Config:
        from_attributes = True


class ScriptLibraryListResponse(BaseModel):
    items: List[ScriptLibrary]
    total: int
    page: int
    page_size: int


class ScriptTag(BaseModel):
    id: str
    category: str
    name: str
    name_en: Optional[str] = None
    description: Optional[str] = None
    examples: List[str] = []
    parent_id: Optional[str] = None


class ScriptTagsByCategory(BaseModel):
    emotion_tones: List[ScriptTag] = []
    relation_types: List[ScriptTag] = []
    contrast_types: List[ScriptTag] = []
    eras: List[ScriptTag] = []
    professions: List[ScriptTag] = []
    gender_targets: List[ScriptTag] = []
    character_genders: List[ScriptTag] = []
    lengths: List[ScriptTag] = []
    age_ratings: List[ScriptTag] = []
